sklearn/metrics/pairwise.py

```diff
@@ -256,6 +289,140 @@ def euclidean_distances(X, Y=None, Y_norm_squared=None, squared=False,
     return distances if squared else np.sqrt(distances, out=distances)
 
 
+def masked_euclidean_distances(X, Y=None, squared=False,
+                               missing_values="NaN", copy=True):
+    """Calculates euclidean distances in the presence of missing values
+
+    Considering the rows of X (and Y=X) as vectors, compute the
+    distance matrix between each pair of vectors. Similarly, if
+    Y is not X, then compute the distance matrix between each
+    pair of vectors (i.e., each row pair) in X and Y.
+
+    This function computes pairwise euclidean distance for vectors
+    in dense matrices X and Y with missing values in arbitrary
+    coordinates.
+
+    The following formula is used for this:
+
+        dist(X, Y) = (X.shape[1] / ((dot(NX, NYT)))) *
+                    (dot((X * X), NYT) - 2 * (dot(X, Y.T)) +
+                    dot(NX, (Y.T * Y.T)))
+
+    where NX and NYT represent the logical-not of the missing masks of
+    X and Y.T, respectively.
+    Formula in matrix form derived by:
+    Shreya Bhattarai <shreya.bhattarai@gmail.com>
```

Comment:
I don't think the docstring is an appropriate place to place credits about implementation details. If you wish, note it in a comment in the code.